### PR TITLE
arch: Merge up_arch.h into up_internal.h

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Impact
Code refactor

## Testing
Pass CI, please ignore the mix case warning.

### CODE REVIEW

File: arch/arm/src/armv7-m/mpu.h [hunk 1]
```diff
@@ -33,8 +33,6 @@
 #  include <stdbool.h>
 #  include <assert.h>
 #  include <debug.h>
-
-#  include "arm_arch.h"
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/mpu.h:37]:
```suggestion
#  include "arm_internal.h"
```

File: arch/arm/src/common/arm_internal.h [hunk 2]
```diff
@@ -182,6 +182,19 @@
...
+#define getreg16(a)    (*(volatile uint16_t *)(a))
+#define putreg16(v,a)  (*(volatile uint16_t *)(a) = (v))
+#define getreg32(a)    (*(volatile uint32_t *)(a))
+#define putreg32(v,a)  (*(volatile uint32_t *)(a) = (v))
+
+/* Non-atomic, but more effective modification of registers */
+
+#define modreg8(v,m,a)  putreg8((getreg8(a) & ~(m)) | ((v) & (m)), a)
+#define modreg16(v,m,a) putreg16((getreg16(a) & ~(m)) | ((v) & (m)), a)
+#define modreg32(v,m,a) putreg32((getreg32(a) & ~(m)) | ((v) & (m)), a)
```
Comment (pkarashchenko) [inline arch/arm/src/common/arm_internal.h:196]:
```suggestion
#define modreg8(v,m,a)  putreg8((getreg8(a) & ~(m)) | ((v) & (m)), (a))
#define modreg16(v,m,a) putreg16((getreg16(a) & ~(m)) | ((v) & (m)), (a))
#define modreg32(v,m,a) putreg32((getreg32(a) & ~(m)) | ((v) & (m)), (a))
```

File: arch/arm/src/dm320/dm320_gio.h [hunk 3]
```diff
@@ -27,7 +27,7 @@
 
 #ifndef __ASSEMBLY__
 # include <stdint.h>
-# include "arm_arch.h"
+# include "arm_internal.h"
 #endif
```
Comment (pkarashchenko) [inline arch/arm/src/dm320/dm320_gio.h:31]:
```suggestion
#endif

#include "arm_internal.h"
```

File: arch/arm/src/armv7-m/mpu.h (hunk 1 above)
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/mpu.h:37]: Do we need to change to `arm_internal.h`?

**Review by pkarashchenko — COMMENTED** [review]
looks like `USB_DmaDesc` has to be `g_usbddesc`

File: arch/arm/src/lpc54xx/lpc54_emc.c [hunk 4]
```diff
@@ -388,7 +387,7 @@ void lpc54_emc_sdram_initialize(
 
   putreg32(2 * EMC_REFRESH_CLOCK_SCALE, LPC54_EMC_DYNREFRESH);
 
-  for (i = 0; i < EMC_SDRAM_WAIT_CYCLES / 2; i ++)
+  for (i = 0; i < EMC_SDRAM_WAIT_CYCLES / 2; i++)
```
Comment (pkarashchenko) [inline arch/arm/src/lpc54xx/lpc54_emc.c:390]:
```suggestion
  for (j = 0; j < EMC_SDRAM_WAIT_CYCLES / 2; j++)
```

File: arch/arm/src/common/arm_internal.h (hunk 2 above)
Comment (xiaoxiang781216) [reply arch/arm/src/common/arm_internal.h:196]: Done.

File: arch/arm/src/dm320/dm320_gio.h (hunk 3 above)
Comment (xiaoxiang781216) [reply arch/arm/src/dm320/dm320_gio.h:31]: Done.

File: arch/arm/src/armv7-m/mpu.h (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/arm/src/armv7-m/mpu.h:37]: Don't, since mpu.h doesn't use putregx/getregx/modifyregx.

File: arch/arm/src/lpc54xx/lpc54_emc.c (hunk 4 above)
Comment (xiaoxiang781216) [reply arch/arm/src/lpc54xx/lpc54_emc.c:390]: Done.

**Comment by xiaoxiang781216** [comment]
> looks like `USB_DmaDesc` has to be `g_usbddesc`

Done.

File: arch/risc-v/src/bl602/bl602_systemreset.c [hunk 5]
```diff
@@ -68,6 +67,7 @@ void up_systemreset(void)
   asm volatile("csrci mstatus, 8");
 
   bl602_romapi_reset_system();
+  for (; ; );
```
Comment (pkarashchenko) [inline arch/risc-v/src/bl602/bl602_systemreset.c:70]:
```suggestion

  /* Wait for the reset */

  for (; ; );
```

File: arch/risc-v/src/mpfs/mpfs_opensbi.c [hunk 6]
```diff
@@ -22,25 +22,12 @@
...
 
-/* OpenSBI will also define NULL. Undefine NULL in order to avoid warning:
- * 'warning: "NULL" redefined'
- */
-
-#ifdef NULL
-  #undef NULL
-#endif
-
+#include <sbi/riscv_io.h>
```
Comment (pkarashchenko) [inline arch/risc-v/src/mpfs/mpfs_opensbi.c:30]: Probably after this PR I will try to renew my changes with OpenSBI NuttX types header file.

File: arch/sparc/src/common/up_internal.h [hunk 7]
```diff
@@ -175,6 +182,12 @@ extern uint32_t _bmxdupba_address;   /* BMX register setting */
  * functions prototyped in include/nuttx/arch.h.
  */
 
+/* Atomic modification of registers */
+
+void modifyreg8(unsigned int addr, uint8_t clearbits, uint8_t setbits);
+void modifyreg16(unsigned int addr, uint16_t clearbits, uint16_t setbits);
+void modifyreg32(unsigned int addr, uint32_t clearbits, uint32_t setbits);
+
```
Comment (pkarashchenko) [inline arch/sparc/src/common/up_internal.h:190]: Can we move this block before `/* Common Functions *********************************************************/`, just for consistency with other similar changes?

File: arch/z80/src/common/z80_internal.h [hunk 8]
```diff
@@ -73,6 +73,21 @@
...
+/* Register access macros ***************************************************
+ *
+ * The register access mechanism provided in ez8.h differs from the useful in
+ * other NuttX architectures. The following NuttX common macros will at least
+ * make the access compatible at the source level (however, strict type check
+ * is lost).
+ */
+
+#define getreg8(a)          (a)
+#define putreg8(v,a)        (a) = (v)
```
Comment (pkarashchenko) [inline arch/z80/src/common/z80_internal.h:85]:
```suggestion
#define putreg8(v,a)        ((a) = (v))
```

File: arch/z80/src/common/z80_internal.h [hunk 9]
```diff
@@ -73,6 +73,21 @@
...
+#define getreg16(a)         (a)
+#define putreg16(v,a)       (a) = (v)
```
Comment (pkarashchenko) [inline arch/z80/src/common/z80_internal.h:87]:
```suggestion
#define putreg16(v,a)       ((a) = (v))
```

File: arch/z80/src/common/z80_internal.h [hunk 10]
```diff
@@ -73,6 +73,21 @@
...
+#define getreg32(a)         (a)
+#define putreg32(v,a)       (a) = (v)
```
Comment (pkarashchenko) [inline arch/z80/src/common/z80_internal.h:89]:
```suggestion
#define putreg32(v,a)       ((a) = (v))
```

File: boards/arm/sama5/sama5d4-ek/src/sam_sdram.c [hunk 11]
```diff
@@ -274,7 +273,7 @@ static void sam_config_slaveddr(void)
  *   "Two DDR2/SDRAM (MT47H64M16HR) used as main system memory (256 MByte).
  *   The board includes 2 Gbits of on-board solderedDDR2 (double data rate)
  *   SDRAM. The footprints can also host two DDR2 (MT47H128M16RT) from
- *   Micron� for a total of 512 MBytes of DDR2 memory. The memory bus is 32
+ *   Micron� for a total of 512 MBytes of DDR2 memory. The memory bus is 32
```
Comment (pkarashchenko) [inline boards/arm/sama5/sama5d4-ek/src/sam_sdram.c:276]: Please revert or replace with something meaningful

Comment (xiaoxiang781216) [reply boards/arm/sama5/sama5d4-ek/src/sam_sdram.c:276]: Done.

File: arch/z80/src/common/z80_internal.h (hunk 10 above)
Comment (xiaoxiang781216) [reply arch/z80/src/common/z80_internal.h:89]: Done.

File: arch/z80/src/common/z80_internal.h (hunk 9 above)
Comment (xiaoxiang781216) [reply arch/z80/src/common/z80_internal.h:87]: Done.

File: arch/z80/src/common/z80_internal.h (hunk 8 above)
Comment (xiaoxiang781216) [reply arch/z80/src/common/z80_internal.h:85]: Done.

File: arch/risc-v/src/bl602/bl602_systemreset.c (hunk 5 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/bl602/bl602_systemreset.c:70]: Done.

File: arch/sparc/src/common/up_internal.h (hunk 7 above)
Comment (xiaoxiang781216) [reply arch/sparc/src/common/up_internal.h:190]: But `/* Common Functions *********************************************************/` is before this block in all other place:
 https://github.com/apache/incubator-nuttx/pull/5720/files?file-filters%5B%5D=.h&show-deleted-files=true&show-viewed-files=true#diff-206bf33ecfcef35eb3b5bf095ebc866558dc44a86ec3bfcf099295c304c944e3R181-R193
https://github.com/apache/incubator-nuttx/blob/master/arch/xtensa/src/common/xtensa.h#L217-L228

Comment (pkarashchenko) [reply arch/sparc/src/common/up_internal.h:190]: Ok

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Comment by hartmannathan** [comment]
Applied the "breaking change" label as this requires out-of-tree boards that `#include "arm_arch.h"` to remove that include (and `#include "arm_internal.h"` if necessary). I'll add an entry to the draft release notes at: [https://cwiki.apache.org/confluence/display/NUTTX/NuttX+10.3](https://cwiki.apache.org/confluence/display/NUTTX/NuttX+10.3). Cheers

**Comment by pkarashchenko** [comment]
@hartmannathan thank you!

**Pull request merged**